src/test/java/libai/nn/supervised/MLPTest.java

```diff
@@ -235,7 +237,160 @@ public void testDemo() {
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testWrongBeta() {
+		new MLP(new int[2], new Function[2], -1);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testWrongBeta2() {
+		new MLP(new int[2], new Function[2], 2);
+	}
+
+	@Test
+	public void testWithProgressBarStandard() {
+		MLP mlp = new MLP(
+			new int[]{2, 16, 1},
+			new Function[]{
+				new Identity(),
+				new Sinc(),
+				new Identity()
+			}
+		);
+		mlp.setProgressBar(progress);
+		mlp.setTrainingType(MLP.STANDARD_BACKPROPAGATION);
+		Matrix[] ins = new Matrix[4];
+		ins[0] = new Matrix(2, 1, new double[]{0, 0});
+		ins[1] = new Matrix(2, 1, new double[]{0, 1});
+		ins[2] = new Matrix(2, 1, new double[]{1, 0});
+		ins[3] = new Matrix(2, 1, new double[]{1, 1});
+		Matrix[] out = new Matrix[4];
+		out[0] = new Matrix(1, 1, new double[]{1});
+		out[1] = new Matrix(1, 1, new double[]{0});
+		out[2] = new Matrix(1, 1, new double[]{0});
+		out[3] = new Matrix(1, 1, new double[]{0});
+		mlp.train(ins, out, 0.01, 1000000, 0, 4, 0.01);
+		assumeTrue("MLP didn't converge, try again", 0.01 > mlp.error(ins, out));
+		Matrix res = new Matrix(1, 1);
+		mlp.simulate(ins[0], res);
+		assertEquals(1, round(res.position(0, 0)));
+		mlp.simulate(ins[1], res);
+		assertEquals(0, round(res.position(0, 0)));
+		mlp.simulate(ins[2], res);
+		assertEquals(0, round(res.position(0, 0)));
+		mlp.simulate(ins[3], res);
+		assertEquals(0, round(res.position(0, 0)));
+	}
+
+	@Test
+	public void testWithProgressBarResilient() {
+		MLP mlp = new MLP(
+			new int[]{2, 16, 1},
+			new Function[]{
+				new Identity(),
+				new Sinc(),
+				new Identity()
+			}
+		);
+		mlp.setProgressBar(progress);
+		mlp.setTrainingType(MLP.RESILENT_BACKPROPAGATION);
+		Matrix[] ins = new Matrix[4];
+		ins[0] = new Matrix(2, 1, new double[]{0, 0});
+		ins[1] = new Matrix(2, 1, new double[]{0, 1});
+		ins[2] = new Matrix(2, 1, new double[]{1, 0});
+		ins[3] = new Matrix(2, 1, new double[]{1, 1});
+		Matrix[] out = new Matrix[4];
+		out[0] = new Matrix(1, 1, new double[]{1});
+		out[1] = new Matrix(1, 1, new double[]{0});
+		out[2] = new Matrix(1, 1, new double[]{0});
+		out[3] = new Matrix(1, 1, new double[]{0});
+		mlp.train(ins, out, 0.01, 1000000, 0, 4, 0.01);
+		assumeTrue("MLP didn't converge, try again", 0.01 > mlp.error(ins, out));
+		Matrix res = new Matrix(1, 1);
+		mlp.simulate(ins[0], res);
+		assertEquals(1, round(res.position(0, 0)));
+		mlp.simulate(ins[1], res);
+		assertEquals(0, round(res.position(0, 0)));
+		mlp.simulate(ins[2], res);
+		assertEquals(0, round(res.position(0, 0)));
+		mlp.simulate(ins[3], res);
+		assertEquals(0, round(res.position(0, 0)));
+	}
+
+	@Test
+	public void testWithProgressBarMomentum() {
+		MLP mlp = new MLP(
+			new int[]{2, 16, 1},
+			new Function[]{
+				new Identity(),
+				new Sinc(),
+				new Identity()
+			}
+		);
+		mlp.setProgressBar(progress);
+		mlp.setTrainingType(MLP.MOMEMTUM_BACKPROPAGATION, 0.5);
```

Comment:
Off topic of this PR. I have been thinking about the way of setting the training algorithm of the MLP, and i think we could create an abstraction around it (and interface) and pass an instance of the training algorithm on the creation time, in the same way we are passing the array of functions for the layers or the way the kernel type is passed on the SVM. In this way it will cleanup the code of the MLP delegating the implementation of the backpropagation to another class and will be easier to create new implementations that receive more parameters (as discussed in one of the issues)